encodings/fastlanes/src/bitpacking/compute/search_sorted.rs

```diff
@@ -0,0 +1,94 @@
+use std::cmp::Ordering;
+
+use fastlanes::BitPacking;
+use vortex::compute::search_sorted::{
+    search_sorted, IndexOrd, Len, SearchResult, SearchSorted, SearchSortedFn, SearchSortedSide,
+};
+use vortex::ArrayDType;
+use vortex_dtype::{match_each_unsigned_integer_ptype, NativePType};
+use vortex_error::VortexResult;
+use vortex_scalar::Scalar;
+
+use crate::{unpack_single_primitive, BitPackedArray};
+
+impl SearchSortedFn for BitPackedArray {
+    fn search_sorted(&self, value: &Scalar, side: SearchSortedSide) -> VortexResult<SearchResult> {
+        let ptype = self.ptype();
+        match_each_unsigned_integer_ptype!(ptype, |$P| {
+            let unwrapped_value: $P = value.cast(self.dtype())?.try_into().unwrap();
+            if let Some(patches_array) = self.patches() {
+                if (unwrapped_value.leading_zeros() as usize) < ptype.bit_width() - self.bit_width() {
+                    search_sorted(&patches_array, value.clone(), side)
+                } else {
+                    Ok(SearchSorted::search_sorted(&BitPackedSearch(self), &unwrapped_value, side))
+                }
+            } else {
+                Ok(SearchSorted::search_sorted(&BitPackedSearch(self), &unwrapped_value, side))
+            }
+        })
+    }
+}
+
+/// This wrapper exists, so that you can't invoke SearchSorted::search_sorted directly on BitPackedArray as it omits searching patches
+#[derive(Debug)]
+struct BitPackedSearch<'a>(&'a BitPackedArray);
+
+impl<T: BitPacking + NativePType> IndexOrd<T> for BitPackedSearch<'_> {
+    fn index_cmp(&self, idx: usize, elem: &T) -> Option<Ordering> {
+        // SAFETY: Used in search_sorted_by which ensures that idx is within bounds
+        let val: T = unsafe {
+            unpack_single_primitive(
+                self.0
+                    .packed()
+                    .flatten_primitive()
```

Comment:
Not sure you should be doing this inside the index_cmp? Can we do it once outside, e.g. when you construct BitPackedSearch?

Reply:
Sure